File: app/src/main/java/com/dougkeen/bart/presentation/DepartureNotificationFactory.java
package com.dougkeen.bart.presentation;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;

import androidx.core.app.NotificationCompat;

import com.dougkeen.bart.R;
import com.dougkeen.bart.activities.TripInProgressActivity;
import com.dougkeen.bart.model.Constants;
import com.dougkeen.bart.model.Departure;
import com.dougkeen.bart.services.BoardedDepartureService;

import java.util.Locale;

/** Builds the foreground notification for a followed departure. */
public final class DepartureNotificationFactory {
    private DepartureNotificationFactory() {
    }

    public static Notification create(Context context, Departure departure) {
        final int halfMinutes = (departure.getMeanSecondsLeft() + 15) / 30;
        float minutes = halfMinutes / 2f;
        final String minutesText = (minutes < 1) ? "Less than one minute"
                : (String.format(Locale.US, "~%.1f minute", minutes)
                + ((minutes != 1.0) ? "s" : ""));
        final String directionText = departure.getOrigin().shortName + " to "
                + departure.getPassengerDestination().shortName;

        Intent cancelAlarmIntent = new Intent(context, BoardedDepartureService.class);
        cancelAlarmIntent.putExtra("cancelNotifications", true);

        String channelId = context.getString(R.string.notification_channel_id);
        NotificationCompat.Builder notificationBuilder =
                new NotificationCompat.Builder(context, channelId)
                        .setSmallIcon(R.drawable.ic_stat_notification)
                        .setContentTitle(minutesText + " until departure")
                        .setContentIntent(notificationIntent(context))
                        .setDeleteIntent(deleteNotificationIntent(context));

        if (departure.getMeanSecondsLeft() > 0) {
            notificationBuilder.setWhen(System.currentTimeMillis()
                    + departure.getMeanSecondsLeft() * 1000L)
                    .setUsesChronometer(true);
        }

        notificationBuilder.setContentText(directionText);
        if (departure.isAlarmPending()) {
            PendingIntent pendingIntent = PendingIntent.getService(
                    context, 0, cancelAlarmIntent, PendingIntent.FLAG_IMMUTABLE);
            String subText = "Alarm " + departure.getAlarmLeadTimeMinutes()
                    + " minutes before departure";
            notificationBuilder
                    .addAction(R.drawable.ic_action_cancel_alarm,
                            "Cancel alarm", pendingIntent)
                    .setSubText(subText);
        }

        return notificationBuilder.build();
    }

    private static PendingIntent notificationIntent(Context context) {
        Intent targetIntent = new Intent(context, TripInProgressActivity.class);
        targetIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP
                | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        return PendingIntent.getActivity(context, 0, targetIntent,
                PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT);
    }

    private static PendingIntent deleteNotificationIntent(Context context) {
        Intent targetIntent = new Intent(context, BoardedDepartureService.class);
        targetIntent.putExtra(Constants.CLEAR_DEPARTURE, true);
        return PendingIntent.getService(context, 0, targetIntent,
                PendingIntent.FLAG_IMMUTABLE);
    }
}

File: app/src/main/kotlin/in/izyum/bart/presentation/DepartureNotificationFactory.kt
package `in`.izyum.bart.presentation

import android.app.Notification
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.app.NotificationChannel
import android.app.NotificationManager
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import `in`.izyum.bart.R
import `in`.izyum.bart.activities.RouteArguments
import `in`.izyum.bart.activities.TripInProgressActivity
import `in`.izyum.bart.model.Departure
import `in`.izyum.bart.model.TimeSource
import `in`.izyum.bart.platform.DepartureAlarmScheduler
import `in`.izyum.bart.receivers.DepartureTrackingReceiver

/** Builds the foreground notification used while a departure alarm is pending. */
object DepartureNotificationFactory {
    @JvmStatic
    fun create(
        context: Context,
        departure: Departure,
        alarmScheduler: DepartureAlarmScheduler?,
        timeSource: TimeSource,
    ): Notification {
        val nowMillis = timeSource.nowMillis()
        val secondsLeft = departure.getMeanSecondsLeft(
            departure.minEstimate,
            departure.maxEstimate,
            nowMillis,
        )
        val minutes = (secondsLeft + 15) / 30 / 2f
        val minutesText = when {
            minutes < 1 -> context.getString(R.string.notification_less_than_minute)
            minutes == 1f -> context.getString(R.string.notification_minutes_until_departure, minutes)
            else -> context.getString(R.string.notification_minutes_until_departures, minutes)
        }
        val directionText = context.getString(
            R.string.notification_direction,
            departure.origin?.shortName.orEmpty(),
            (departure.passengerDestination ?: departure.trainDestination)?.shortName.orEmpty(),
        )
        val cancelAlarmIntent = Intent(context, DepartureTrackingReceiver::class.java)
            .setAction(DepartureTrackingReceiver.ACTION_CANCEL_ALARM)
        val channelId = context.getString(R.string.notification_channel_id)
        val builder = NotificationCompat.Builder(context, channelId)
            .setSmallIcon(R.drawable.ic_stat_notification)
            .setContentTitle(minutesText)
            .setContentIntent(notificationIntent(context, departure))
            .setDeleteIntent(deleteNotificationIntent(context))
            .setContentText(directionText)
            .setOnlyAlertOnce(true)

        if (secondsLeft > 0) {
            builder
                .setWhen(nowMillis + secondsLeft * 1000L)
                .setUsesChronometer(true)
        }
        if (alarmScheduler?.isPending == true) {
            val pendingIntent = PendingIntent.getBroadcast(
                context,
                0,
                cancelAlarmIntent,
                PendingIntent.FLAG_IMMUTABLE,
            )
            builder
                .addAction(
                    R.drawable.ic_action_cancel_alarm,
                    context.getString(R.string.notification_cancel_alarm),
                    pendingIntent,
                )
                .setSubText(context.resources.getQuantityString(
                    R.plurals.notification_alarm,
                    alarmScheduler.leadTimeMinutes,
                    alarmScheduler.leadTimeMinutes,
                ))
        }
        return builder.build()
    }

    private fun notificationIntent(context: Context, departure: Departure): PendingIntent {
        val targetIntent = Intent(context, TripInProgressActivity::class.java).apply {
            addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP)
        }
        RouteArguments.putTrip(
            targetIntent,
            departure.getStationPair(),
            departure.identity,
            RouteArguments.MODE_FOLLOWED,
        )
        return PendingIntent.getActivity(
            context,
            0,
            targetIntent,
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT,
        )
    }

    private fun deleteNotificationIntent(context: Context): PendingIntent {
        val targetIntent = Intent(context, DepartureTrackingReceiver::class.java)
            .setAction(DepartureTrackingReceiver.ACTION_CLEAR_DEPARTURE)
        return PendingIntent.getBroadcast(context, 0, targetIntent, PendingIntent.FLAG_IMMUTABLE)
    }

    fun show(
        context: Context,
        departure: Departure,
        repository: `in`.izyum.bart.data.FollowedTripRepository,
        timeSource: TimeSource,
    ) {
        context.getSystemService(NotificationManager::class.java).createNotificationChannel(
            NotificationChannel(
                context.getString(R.string.notification_channel_id),
                context.getString(R.string.notification_channel_name),
                NotificationManager.IMPORTANCE_DEFAULT,
            ).apply {
                enableLights(false)
                enableVibration(false)
                setSound(null, null)
            },
        )
        try {
            NotificationManagerCompat.from(context).notify(
                123, create(context, departure, repository.getAlarmScheduler(), timeSource),
            )
        } catch (_: SecurityException) {
            // Notification permission can be revoked independently of alarms.
        }
    }

    fun cancel(context: Context) {
        NotificationManagerCompat.from(context).cancel(123)
    }
}
